### ISSUE TITLE
Optional relation forced

### ISSUE THREAD

**Issue by Lisciowsky** [aws-amplify/amplify-category-api#3077]
### How did you install the Amplify CLI?

npm install -g @aws-amplify/cli

### If applicable, what version of Node.js are you using?

v18.20.3

### Amplify CLI Version

12.13.1

### What operating system are you using?

ubuntu 22.04

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

no manual changes 

### Describe the bug
```graphql
type File
  @model
{
  id: ID!
  ownerEmail: String!
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  fileName: String!
  filePath: String!
  storeID: ID
  store: Store @belongsTo(fields: ["storeID"])
}
```

```js
createFile = """
  mutation CreateFile(
    $input: CreateFileInput!
    $condition: ModelFileConditionInput
  ) {
    createFile(input: $input, condition: $condition) {
      id
      ownerEmail
      createdAt
      updatedAt
      fileName
      filePath
      storeID
      store {
        id
        ownerEmail
        createdAt
        updatedAt
        name
        __typename
      }
      __typename
    }
  }
"""
```

createFile mutation variables:

"fileName": "text_file",
"filePath": "/random_user/text_file",
"ownerEmail": "test@o2.pl",
"storeID": "3afa26ba-b30f-405e-aee6-804a9a9c19a6"
    
With the newest version of amplify it is not possible to resolve the Store, I had to dump the version of the amplify cli so that the created resolvers would understand that store can be associated with a file but doesn't have to and if the store id is given during the file creation mutation it then can resolve the actual store.
    
You have a serious problem with hasMany and belongs to directive, no doubt about that.

### Expected behavior

I expected to have successful graphql call during file creation mutation with or without providing the existing associated store id with the newest version of amplify cli. This isn't the case

### Reproduction steps

If you would like to reproduce the bug just install the newest aws cli and then create two models 

one model will have 
storeID: ID
  store: Store @belongsTo(fields: ["storeID"])
  
 another model will have:
 files: [File] @hasMany(fields: ["id"])
 
 And try to create the file without specifying the store in this example, or omitting it completely.
 
 
I am forced to use the older version 12.1.0 which created the "reasonable" resolvers.

### Log output

<details>

```
# Put your logs below this line


```

</details>

### Additional information

N/A

### Before submitting, please confirm:

- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

**Comment by Siqi-Shan** [comment]
Hi @Lisciowsky, thanks for raising the issue! We'll investigate and see what's going on, and get back to you once we have a solution.

**Comment by chrisbonifacio** [comment]
Hi @Lisciowsky 👋 A few questions to better understand the issue described.

1. How is the relationship being forced? Can you provide any error messages or logs you've received related to this issue?
2. The `createFile` mutation returns `store` and possibly required fields on store so if one doesn't exist, those fields being included in the selection set would likely cause an error.
3. Can you confirm if there's a difference in the resolver logic or in the mutation selection set between the working and non-working CLI versions?
4. Can you share the schema for both `File` and `Storage` so can reproduce the issue as closely to your setup as possible? It might help identify any issues related to non-obvious things like auth rules as well.

in the meantime, as a potential workaround, you could try omitting the entire `store`, or at least any required fields on it from the selection set.

i'm not sure what the non-nullable fields on `store` are, but usually `id`, `createdAt`, `updatedAt` are non-nullable so I removed them from the mutation selection set below. 

```graphql
createFile(input: $input, condition: $condition) {
      id
      ownerEmail
      createdAt
      updatedAt
      fileName
      filePath
      storeID
      store {
        ownerEmail
        name
        __typename
      }
      __typename
    }
  }
```

**Comment by Lisciowsky** [comment]
Hello,

- **_How is the relationship being forced? Can you provide any error messages or logs you've received related to this issue?_**
- _**The createFile mutation returns store and possibly required fields on store so if one doesn't exist, those fields being included in the selection set would likely cause an error.**_

It is forced because the createFile mutation does include the [**optional**] store, and without providing it, it causes the error as you have mentioned.

- **_Can you confirm that there's a difference in the resolver logic or in the mutation selection set between the working and non-working CLI versions?_**

Yes I can confirm, I had to compare on 2 separate environemnts DEV (created with newest version of amplify cli) and STAGE with (12.1.0)

DEV:
```
#if( $ctx.stash.deniedField )
  #return($util.toJson(null))
#end

#set( $partitionKeyValue = $util.defaultIfNull($ctx.stash.connectionAttributes.get("officeProjectsId"), $ctx.source.officeProjectsId) )

#if( $util.isNullOrEmpty($partitionKeyValue) )
  #return($util.toJson(null))
#else
  #set( $GetRequest = {
    "version": "2018-05-29",
    "operation": "Query",
    "query": {
      "expression": "#partitionKey = :partitionValue",
      "expressionNames": {
        "#partitionKey": "id"
      },
      "expressionValues": {
        ":partitionValue": $util.dynamodb.toDynamoDB($partitionKeyValue)
      }
    }
  })

  #if( !$util.isNullOrEmpty($ctx.stash.authFilter) )
    $util.qr($GetRequest.put("filter", $util.transform.toDynamoDBFilterExpression($ctx.stash.authFilter)))
  #end

  $util.toJson($GetRequest)
#end

#if( $ctx.error )
$util.error($ctx.error.message, $ctx.error.type)
#else
  #if( !$ctx.result.items.isEmpty() && $ctx.result.scannedCount == 1 )
    #set( $resultValue = $ctx.result.items[0] )
    #set( $operation = $util.defaultIfNull($ctx.source.get("__operation"), null) )
    #if( $operation == "Mutation" )
      $util.qr($resultValue.put("__operation", "Mutation"))
    #end
    $util.toJson($resultValue)
  #else
    #if( $ctx.result.items.isEmpty() && $ctx.result.scannedCount == 1 )
$util.unauthorized()
    #end
    $util.toJson(null)
  #end
#end
```


STAGE:
```
#if( $ctx.stash.deniedField )
  #return($util.toJson(null))
#end
#set( $partitionKeyValue = $util.defaultIfNull($ctx.stash.connectionAttibutes.get("officeProjectsId"), $ctx.source.officeProjectsId) )
#if( $util.isNull($partitionKeyValue) )
  #return
#else
  #set( $GetRequest = {
  "version": "2018-05-29",
  "operation": "Query"
} )
  $util.qr($GetRequest.put("query", {
  "expression": "#partitionKey = :partitionValue",
  "expressionNames": {
      "#partitionKey": "id"
  },
  "expressionValues": {
      ":partitionValue": $util.parseJson($util.dynamodb.toDynamoDBJson($util.defaultIfNullOrBlank($partitionKeyValue, "___xamznone____")))
  }
}))
  #if( !$util.isNullOrEmpty($ctx.stash.authFilter) )
    $util.qr($GetRequest.put("filter", $util.parseJson($util.transform.toDynamoDBFilterExpression($ctx.stash.authFilter))))
  #end
  $util.toJson($GetRequest)
#end

#if( $ctx.error )
$util.error($ctx.error.message, $ctx.error.type)
#else
  #if( !$ctx.result.items.isEmpty() && $ctx.result.scannedCount == 1 )
    $util.toJson($ctx.result.items[0])
  #else
    #if( $ctx.result.items.isEmpty() && $ctx.result.scannedCount == 1 )
$util.unauthorized()
    #end
    $util.toJson(null)
  #end
#end
```

This is CGPT comment:
```
Why Stage Works, but Dev Doesn’t:
Different Resolver Logic: The Stage resolver uses toDynamoDBJson($util.defaultIfNullOrBlank(...)) while the Dev resolver directly uses $util.dynamodb.toDynamoDB($partitionKeyValue). If the officeProjectsId is blank, the Stage resolver gracefully handles it with "___xamznone____", whereas the Dev resolver fails.

Error Handling Differences: The Stage resolver is more concise and handles error cases better, directly returning unauthorized() or null, while the Dev resolver introduces unnecessary complexity in logic and error handling. 
```

**_Can you share the schema for both File and Storage so we can reproduce the issue as closely to your setup as possible? It might help identify any issues related to non-obvious things like auth rules as well._**

You can replicate this one easily as long as you have one to many with @belongsTo ans @hasMany directives on the one-to-many model relation.

Here is the example grapqhl schema definition:
```
type Store
  @model
  @auth(
    rules: [
      { allow: owner, ownerField: "ownerEmail", identityClaim: "email", provider: oidc, operations: [create, update, delete, read] }
      { allow: groups, groups: ["Admins"], groupClaim: "groups", provider: oidc, operations: [create, update, delete, read] }
    ]
  ) {
  id: ID!
  ownerEmail: String!
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  name: String!
  files: [File] @hasMany
}

type File
  @model
  @auth(
    rules: [
      { allow: owner, ownerField: "ownerEmail", identityClaim: "email", provider: oidc, operations: [create, update, delete, read] }
      { allow: groups, groups: ["Admins"], groupClaim: "groups", provider: oidc, operations: [create, update, delete, read] }
    ]
  ) {
  id: ID!
  ownerEmail: String!
  createdAt: AWSDateTime
  updatedAt: AWSDateTime
  fileName: String!
  filePath: String!
  storeID: ID
  store: Store @belongsTo(fields: ["storeID"])
}

```

Just to summarize, the newest version of the aws cli does create the resolvers but they force the File (following the example schema above) to ensure it points to the store. Your tempfix would certainly work (not including the store in the grapqhl).

 I **do not want to modify the auto-generated amplify schema** and I believe nobody wants to every time I code gen or make some changes. Then I would end up in a scenario where a file without an associated store would have to use a different query, that's just bad news.


PS: The resolver differences on DEV & STAGE example were taken from different project but the issue is exactly the same.

I hope it helps you to fix this, or perhaps roll back the changes that impact the resolver creation

**Comment by armanian** [comment]
I think, I have same Issue. My schema:

```GRAPHQL
type Chat
  @model(queries: null, mutations: { create: "createChat", update: "updateChat" })
  @auth(rules: [{ allow: owner, ownerField: "owners", operations: [create, read, update] }]) {
  id: ID!
  lastMessageID: ID
  owners: [ID!]! @auth(rules: [{ allow: owner, ownerField: "owners", operations: [create, read] }])
  updatedAt: AWSDateTime!

  LastMessage: ChatMessage @hasOne(fields: ["lastMessageID"])
}

type ChatMessage
  @model(queries: null, mutations: { create: "createChatMessage" })
  @auth(rules: [{ allow: owner, ownerField: "owners", operations: [create, read] }]) {
  id: ID!
  chatID: ID
  images: [ID]
  message: String
  owners: [ID!]!
  userID: ID!
  createdAt: AWSDateTime

  Chat: Chat @hasOne(fields: ["chatID"])
}
```
When I use CreateChat I get the following error message:
```JSON
[{"locations": null, "message": "Cannot return null for non-nullable type: 'ID' within parent 'ChatMessage' (/createChat/LastMessage/userID)", "path": [Array]}, {"locations": null, "message": "Cannot return null for non-nullable type: 'null' within parent 'ChatMessage' (/createChat/LastMessage/owners)", "path": [Array]}
```
Error message although the `lastMessageID` field is optional. It used to work like this, but no longer does.

**Comment by armanian** [comment]
@Lisciowsky I found out that if I deploy with version `12.12.1`, everything works correctly. If I deploy with version `12.12.2`, then I get the error message. Can you confirm this?

**Comment by AnilMaktala** [comment]
Hey @armanian Thank you for providing additional details. Could you please try using the latest version of the CLI (12.13.1) and let us know if the issue persists?

**Comment by bobbyu99** [comment]
@Lisciowsky @armanian Thanks for sharing this info. I am working on repro-ing your error.

**Comment by armanian** [comment]
@AnilMaktala yes, with the latest version of CLI (12.13.1) the problem is still present. Downgrade to `12.12.1` works again.

**Comment by AnilMaktala** [comment]
@armanian Thank you for confirming. This appears to be a duplicate of #2609. This behavior is intentional, and you can opt out of the new behavior by setting the `subscriptionsinheritprimaryauth` feature flag to `true`.

For more information, please refer to the updated documentation:

[Build a Backend: GraphQL API - Subscribe to Data](https://docs.amplify.aws/gen1/javascript/build-a-backend/graphqlapi/subscribe-data/)
[CLI Reference: Feature Flags - subscriptionsInheritPrimaryAut](https://docs.amplify.aws/gen1/javascript/tools/cli/reference/feature-flags/#subscriptionsInheritPrimaryAuth)

**Comment by armanian** [comment]
@AnilMaktala Thank you very much, that has solved the problem 🥳
But strange that such a beaking change is integrated in a patch version.

**Comment by bobbyu99** [comment]
@Lisciowsky Please let us know if the [fix](https://github.com/aws-amplify/amplify-category-api/issues/3077#issuecomment-2563762721) above address your issue. Thanks.

**Comment by ivadenis** [comment]
> @AnilMaktala Thank you very much, that has solved the problem 🥳 But strange that such a beaking change is integrated in a patch version.

100%! this broke my production app for no reason! i spent so much time and nerves trying to figure out what's wrong...Engineering excellence at aws is sub-standard...

aws, changes like these are MAJOR! you have to change the major version of the package...

**Comment by ivadenis** [comment]
> @armanian Thank you for confirming. This appears to be a duplicate of #2609. This behavior is intentional, and you can opt out of the new behavior by setting the `subscriptionsinheritprimaryauth` feature flag to `true`.
> 
> For more information, please refer to the updated documentation:
> 
> [Build a Backend: GraphQL API - Subscribe to Data](https://docs.amplify.aws/gen1/javascript/build-a-backend/graphqlapi/subscribe-data/) [CLI Reference: Feature Flags - subscriptionsInheritPrimaryAut](https://docs.amplify.aws/gen1/javascript/tools/cli/reference/feature-flags/#subscriptionsInheritPrimaryAuth)

Will it work for mutations as well?

**Comment by ivadenis** [comment]
> @armanian Thank you for confirming. This appears to be a duplicate of #2609. This behavior is intentional, and you can opt out of the new behavior by setting the `subscriptionsinheritprimaryauth` feature flag to `true`.
> 
> For more information, please refer to the updated documentation:
> 
> [Build a Backend: GraphQL API - Subscribe to Data](https://docs.amplify.aws/gen1/javascript/build-a-backend/graphqlapi/subscribe-data/) [CLI Reference: Feature Flags - subscriptionsInheritPrimaryAut](https://docs.amplify.aws/gen1/javascript/tools/cli/reference/feature-flags/#subscriptionsInheritPrimaryAuth)

is it `subscriptionsInheritPrimaryAuth` as in the doc, or `subscriptionsinheritprimaryauth` per your message? does casing affect anything?

**Comment by ivadenis** [comment]
@aws team, chose your own punishment for introducing this change:

1. The Scroll of Infinite Deprecations: The team must transcribe an endless scroll of deprecated APIs, each line more obsolete than the last, until they learn the true value of versioning.

2. The Gauntlet of Unending Recompiles: They shall be tasked with recompiling a vast codebase, only to find each build fails due to mysterious, undocumented changes.

3. The Labyrinth of Lost Documentation: The team must navigate a maze filled with outdated and cryptic documentation, seeking the elusive path to clarity and proper versioning.

4. The Siege of the Phantom Bugs: They will be besieged by phantom bugs that appear only in production, vanishing when attempts are made to reproduce them locally.

5. The Quest for the Missing Changelog: A quest to find the missing changelog, hidden deep within the archives of forgotten commits, to restore order to the versioning realm.

**Comment by armanian** [comment]
> is it subscriptionsInheritPrimaryAuth as in the doc, or subscriptionsinheritprimaryauth per your message? does casing affect anything?

@ivadenis Yes, the mutation is working again for me. I also spent hours looking for the error in my code.

I have set `subscriptionsInheritPrimaryAuth` in the `cli.json`. It looks like this for me:
```JSON
{
  "features": {
    "graphqltransformer": {
      "addmissingownerfields": true,
      "validatetypenamereservedwords": true,
      "useexperimentalpipelinedtransformer": true,
      "enableiterativegsiupdates": true,
      "secondarykeyasgsi": true,
      "skipoverridemutationinputtypes": true,
      "securityEnhancementNotification": false,
      "showfieldauthnotification": false,
      "transformerversion": 2,
      "suppressschemamigrationprompt": true,
      "subscriptionsInheritPrimaryAuth": true
    },
    "frontend-ios": {
      "enablexcodeintegration": true
    },
    "auth": {
      "enablecaseinsensitivity": true
    },
    "codegen": {
      "useappsyncmodelgenplugin": true,
      "usedocsgeneratorplugin": true,
      "usetypesgeneratorplugin": true,
      "cleangeneratedmodelsdirectory": true,
      "retaincasestyle": true
    },
    "appsync": {
      "generategraphqlpermissions": true
    },
    "project": {
      "overrides": true
    }
  },
  "debug": {
    "shareProjectConfig": false
  }
}
```

**Comment by AnilMaktala** [comment]
Hey 👋 , This issue is being closed due to inactivity. If you are still experiencing the same problem and need further assistance, please feel free to leave a comment. This will enable us to reopen the issue and provide you with the necessary support.
